core/src/main/java/org/infinispan/container/offheap/SegmentedBoundedOffHeapDataContainer.java

```diff
@@ -244,51 +238,42 @@ private void ensureSize() {
 
    private class OffHeapMapSupplier implements Supplier<ConcurrentMap<WrappedBytes,
          InternalCacheEntry<WrappedBytes, WrappedBytes>>> {
-      private final boolean addAllocationSize;
-      private final int allocationSize;
-
-      /**
-       * Supplier that creates a new OffHeapConcurrentMap per invocation
-       * @param addAllocationSize whether starting/stopping map should add the alloction size to count
-       * @param allocationSize rounded up to nearest power of two
-       */
-      private OffHeapMapSupplier(boolean addAllocationSize, int allocationSize) {
-         this.addAllocationSize = addAllocationSize;
-         this.allocationSize = Util.findNextHighestPowerOfTwo(allocationSize);
-      }
-
       @Override
       public ConcurrentMap<WrappedBytes, InternalCacheEntry<WrappedBytes, WrappedBytes>> get() {
-         // OffHeap concurrent map will round allocationSize to nearest power of 2
-         OffHeapConcurrentMap map = new OffHeapConcurrentMap(allocationSize, allocator, offHeapEntryFactory,
-               offHeapListener) {
-            @Override
-            public void stop() {
-               super.stop();
-               if (addAllocationSize) {
-                  lruLock.lock();
-                  try {
-                     currentSize -= UnpooledOffHeapMemoryAllocator.estimateSizeOverhead(allocationSize << 3);
-                  } finally {
-                     lruLock.unlock();
-                  }
-               }
-            }
-         };
-         if (addAllocationSize) {
-            lruLock.lock();
-            try {
-               currentSize += UnpooledOffHeapMemoryAllocator.estimateSizeOverhead(allocationSize << 3);
-            } finally {
-               lruLock.unlock();
-            }
-         }
+         ExpandingOffHeapConcurrentMap map = new ExpandingOffHeapConcurrentMap(allocator, offHeapEntryFactory,
+               offHeapListener);
          map.start();
          return map;
       }
    }
 
-   private class OffHeapListener implements OffHeapConcurrentMap.EntryListener {
+   private class OffHeapListener implements ExpandingOffHeapConcurrentMap.EntryListener {
+
+      @Override
+      public boolean resize(int pointerCount) {
+         if (useCount) {
+            return true;
+         }
+         lruLock.lock();
+         try {
+            long memoryUsed = ((long) pointerCount) << 3;
+            boolean isNegative = memoryUsed < 0;
+            long change = UnpooledOffHeapMemoryAllocator.estimateSizeOverhead(Math.abs(memoryUsed));
+
+            // If the pointers alone would fill the entire memory cache region, don't let it resize
+            if (change >= maxSize) {
```

Comment:
Shouldn't you disallow the resize if `currentSize + change > maxSize`?

Reply:
I debated about this, the problem is you have multiple maps in this container. And currently once this method returns false the resizing of that specific map sets the threshold which will never let it resize again. I didn't want to slam that door shut unless I know for a fact we can't ever resize. I am all for this change if we provide a way for a map to re-evaluate resizing at a later point (other than clear).

But I haven't figured out a good way to do this yet.

Reply:
Good point, we don't want to limit one map to the initial size forever just because segments aren't balanced and other segment has too many entries. But I don't like going over the limit the user set, either...

I think the cleanest option (but not necessarily the easiest to implement) would be to reserve memory for all the segments (not just the ones actually present in the container) every time the largest map is resized. That way, a write to a segment with a small table will never be rejected because the table cannot be resized, and there's no way to use half the memory for a single map's table. WDYT?

Reply:
So what I have done for now is multiple the memory size by the number of total segments when comparing to the maximum size. I am guessing this is more inline with what you were thinking? This should work pretty well I am thinking.

The only real issue would be if you had a lot of entries mapping to the same segment would be slower to do operations upon that map. But having this many segment collisions is frowned upon in other places as well.

Reply:
Yes, that's exactly what I had in mind.

I'm not really worried about one segment having many more entries than the others except when there are few entries. It could happen e.g. if the values are really big compared to the configured memory size, but I'm not sure it's a realistic scenario.